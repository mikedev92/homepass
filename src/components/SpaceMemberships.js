// @flow
import React from 'react';
import { Share } from 'react-native';
import {
  Badge,
  Right,
  Icon,
  Container,
  Header,
  Left,
  Button,
  ListItem,
  Body,
  Content,
  Text,
  Thumbnail,
  Title,
  List,
} from 'native-base';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { compose } from 'recompose';
import { get, capitalize } from 'lodash';
import Swipeout from 'react-native-swipeout';
import { graphqlLoader } from './common';
import { variables } from '../theme';

type Props = {
  navigation: Navigation,
  data: ApolloData<Space>,
  inviteSpaceMember: () => void,
  removeSpaceMember: string => () => void,
};

function SpaceMemberships({ navigation, data, inviteSpaceMember, removeSpaceMember }: Props) {
  const space = data.node;
  if (!space) {
    return <Container backgroundDark />;
  }
  const memberRequests: SpaceMemberRequestEdge[] = get(space, 'memberRequests.edges') || [];
  const members: SpaceMemberEdge[] = get(space, 'members.edges') || [];
  return (
    <Container backgroundDark>
      <Header bordered>
        <Left>
          <Button onPress={() => navigation.goBack()} transparent primary>
            <Icon name="arrow-back" />
          </Button>
        </Left>
        <Body><Title>Membership</Title></Body>
        <Right />
      </Header>
      <Content>
        <ListItem itemHeader />
        <List>
          <ListItem
            icon
            last={memberRequests.length === 0}
            onPress={() => navigation.navigate('SpaceVisibility', { id: space.id })}
          >
            <Body><Text>Visibility</Text></Body>
            <Right><Text>{capitalize(space.visibility)}</Text><Icon name="arrow-forward" /></Right>
          </ListItem>
          {memberRequests.length > 0 &&
            <ListItem
              icon
              last={memberRequests.length !== 0}
              onPress={() => navigation.navigate('SpaceMembershipRequests', { id: space.id })}
            >
              <Body><Text>Membership Requests</Text></Body>
              <Right>
                <Badge><Text>{memberRequests.length}</Text></Badge>
                <Icon name="arrow-forward" />
              </Right>
            </ListItem>}
        </List>
        <ListItem itemHeader>
          <Text caption>MEMBERS</Text>
        </ListItem>
        <List>
          <ListItem
            last={members.length === 0}
            onPress={() => navigation.navigate('SpaceAddMember', { id: space.id })}
          >
            <Text primary>Add Member</Text>
          </ListItem>
          {members.length > 0 &&
            members.map(
              ({ node }, index) =>
                node &&
                node.user &&
                node.spaceMemberId &&
                <Swipeout
                  backgroundColor="transparent"
                  right={[
                    {
                      backgroundColor: variables.btnDangerBg,
                      text: 'Delete',
                      onPress: removeSpaceMember(node.spaceMemberId),
                    },
                  ]}
                  key={node.spaceMemberId}
                >
                  <ListItem avatar last={members.length - 1 === index}>
                    {node.user.imageUrl &&
                      <Left><Thumbnail small source={{ uri: node.user.imageUrl }} /></Left>}
                    <Body><Text>{node.user.name}</Text></Body>
                  </ListItem>
                </Swipeout>,
            )}
        </List>
        <ListItem itemHeader>
          <Text caption>GUESTS</Text>
        </ListItem>
        <List>
          <ListItem last onPress={inviteSpaceMember}>
            <Text primary>Invite Guest</Text>
          </ListItem>
        </List>
        <Content listPadding>
          <Text caption>
            Invite customers as guests to give theme a live mobile experience
          </Text>
        </Content>
      </Content>
    </Container>
  );
}

const SpaceMembershipsFragment = gql`
  fragment SpaceMembershipsFragment on Space {
      id
      spaceId
      visibility
      members {
        edges {
          node {
            id
            spaceMemberId
            user {
              id
              imageUrl
              name
            }
          }
        }
      }
      memberRequests {
        edges {
          node {
            id
          }
        }
      }
  }
`;

const GetSpaceInviteMutation = gql`
  mutation($input: GetSpaceInviteInput!) {
    getSpaceInvite(input: $input) {
      spaceInvite {
        subject
        body
        inviteUrl
      }
    }
  }
`;

const RemoveSpaceMemberMutation = gql`
  mutation($input: RemoveSpaceMemberInput!) {
    removeSpaceMember(input: $input) {
      space {
        ...SpaceMembershipsFragment
      }
    }
  }
  ${SpaceMembershipsFragment}
`;

const SpaceMembershipsQuery = gql`
  query SpaceMembershipsQuery($id: ID!) {
    node(id: $id) {
      ...SpaceMembershipsFragment
    }
  }
  ${SpaceMembershipsFragment}
`;

export default compose(
  graphql(SpaceMembershipsQuery, {
    options: props => ({ variables: { id: props.navigation.state.params.id } }),
  }),
  graphql(RemoveSpaceMemberMutation, {
    props: ({ mutate }) => ({
      removeSpaceMember: spaceMemberId => () =>
        mutate({
          variables: {
            input: { spaceMemberId },
          },
        }),
    }),
  }),
  graphql(GetSpaceInviteMutation, {
    props: ({ mutate, ownProps }) => ({
      inviteSpaceMember: async () => {
        const mutation = mutate({
          variables: {
            input: {
              spaceId: ownProps.data.node.spaceId,
            },
          },
        });
        const res = await mutation;
        const spaceInvite: SpaceInvitationType = get(res, 'data.getSpaceInvite.spaceInvite');
        Share.share({
          message: spaceInvite.body,
          title: spaceInvite.subject,
        });
      },
    }),
  }),
  graphqlLoader,
)(SpaceMemberships);
