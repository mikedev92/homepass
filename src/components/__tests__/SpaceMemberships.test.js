// @flow
import React from 'react';
import { shallow } from 'enzyme';
import SpaceMemberships from '../SpaceMemberships';

const testProps = [
  {
    data: {
      node: {
        visibility: 'OPEN',
        memberRequests: {
          edges: [{}],
        },
        members: {
          edges: [{ node: { user: { name: 'Name', imageUrl: 'Image' } } }],
        },
      },
    },
  },
  {
    data: {
      node: {
        visibility: 'SECRET',
        memberRequests: {
          edges: [],
        },
        members: {
          edges: [{ node: { spaceMemberId: '123', user: { name: 'Name', imageUrl: 'Image' } } }],
        },
      },
    },
  },
  {
    data: {
      node: {
        visibility: 'PRIVATE',
        memberRequests: {
          edges: [{}, {}],
        },
        members: {
          edges: [],
        },
      },
    },
  },
];
describe('<SpaceMemberships />', () => {
  it('matches snapshots', () => {
    testProps.forEach((props) => {
      const baseProps = {
        inviteSpaceMember: jest.fn(),
        removeSpaceMember: jest.fn(),
      };
      const wrapper = shallow(<SpaceMemberships {...baseProps} {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
