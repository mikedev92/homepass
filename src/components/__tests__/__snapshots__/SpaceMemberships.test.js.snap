// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceMemberships /> matches snapshots 1`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)>
      <Styled(Title)>
        Membership
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)>
    <Styled(ListItem)
      itemHeader={true}
    />
    <Styled(List)>
      <Styled(ListItem)
        icon={true}
        last={false}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            Visibility
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Text)>
            Open
          </Styled(Text)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
      <Styled(ListItem)
        icon={true}
        last={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            Membership Requests
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Badge)>
            <Styled(Text)>
              1
            </Styled(Text)>
          </Styled(Badge)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        MEMBERS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={false}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Add Member
        </Styled(Text)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        GUESTS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={true}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Invite Guest
        </Styled(Text)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(Content)
      listPadding={true}
    >
      <Styled(Text)
        caption={true}
      >
        Invite customers as guests to give theme a live mobile experience
      </Styled(Text)>
    </Styled(Content)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<SpaceMemberships /> matches snapshots 2`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)>
      <Styled(Title)>
        Membership
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)>
    <Styled(ListItem)
      itemHeader={true}
    />
    <Styled(List)>
      <Styled(ListItem)
        icon={true}
        last={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            Visibility
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Text)>
            Secret
          </Styled(Text)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        MEMBERS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={false}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Add Member
        </Styled(Text)>
      </Styled(ListItem)>
      <Swipeout
        backgroundColor="transparent"
        disabled={false}
        right={
          Array [
            Object {
              "backgroundColor": "#d9534f",
              "onPress": undefined,
              "text": "Delete",
            },
          ]
        }
        rowID={-1}
        sectionID={-1}
        sensitivity={50}
      >
        <Styled(ListItem)
          avatar={true}
          last={true}
        >
          <Styled(Left)>
            <Styled(Thumbnail)
              small={true}
              source={
                Object {
                  "uri": "Image",
                }
              }
            />
          </Styled(Left)>
          <Styled(Body)>
            <Styled(Text)>
              Name
            </Styled(Text)>
          </Styled(Body)>
        </Styled(ListItem)>
      </Swipeout>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        GUESTS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={true}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Invite Guest
        </Styled(Text)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(Content)
      listPadding={true}
    >
      <Styled(Text)
        caption={true}
      >
        Invite customers as guests to give theme a live mobile experience
      </Styled(Text)>
    </Styled(Content)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<SpaceMemberships /> matches snapshots 3`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)>
      <Styled(Title)>
        Membership
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)>
    <Styled(ListItem)
      itemHeader={true}
    />
    <Styled(List)>
      <Styled(ListItem)
        icon={true}
        last={false}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            Visibility
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Text)>
            Private
          </Styled(Text)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
      <Styled(ListItem)
        icon={true}
        last={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            Membership Requests
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Badge)>
            <Styled(Text)>
              2
            </Styled(Text)>
          </Styled(Badge)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        MEMBERS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={true}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Add Member
        </Styled(Text)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(ListItem)
      itemHeader={true}
    >
      <Styled(Text)
        caption={true}
      >
        GUESTS
      </Styled(Text)>
    </Styled(ListItem)>
    <Styled(List)>
      <Styled(ListItem)
        last={true}
        onPress={[Function]}
      >
        <Styled(Text)
          primary={true}
        >
          Invite Guest
        </Styled(Text)>
      </Styled(ListItem)>
    </Styled(List)>
    <Styled(Content)
      listPadding={true}
    >
      <Styled(Text)
        caption={true}
      >
        Invite customers as guests to give theme a live mobile experience
      </Styled(Text)>
    </Styled(Content)>
  </Styled(Content)>
</Styled(Container)>
`;
